src/core/lib/backoff/backoff.cc

```diff
@@ -20,23 +20,23 @@
 
 #include <grpc/support/useful.h>
 
-void grpc_backoff_init(grpc_backoff *backoff,
-                       grpc_millis initial_connect_timeout, double multiplier,
-                       double jitter, grpc_millis min_timeout_millis,
-                       grpc_millis max_timeout_millis) {
-  backoff->initial_connect_timeout = initial_connect_timeout;
+void grpc_backoff_init(grpc_backoff *backoff, grpc_millis initial_backoff,
+                       double multiplier, double jitter,
+                       grpc_millis min_backoff, grpc_millis max_backoff) {
+  backoff->initial_backoff = initial_backoff;
   backoff->multiplier = multiplier;
   backoff->jitter = jitter;
-  backoff->min_timeout_millis = min_timeout_millis;
-  backoff->max_timeout_millis = max_timeout_millis;
+  backoff->min_backoff = min_backoff;
+  backoff->max_backoff = max_backoff;
   backoff->rng_state = (uint32_t)gpr_now(GPR_CLOCK_REALTIME).tv_nsec;
 }
 
 grpc_millis grpc_backoff_begin(grpc_exec_ctx *exec_ctx, grpc_backoff *backoff) {
-  backoff->current_timeout_millis = backoff->initial_connect_timeout;
-  const grpc_millis first_timeout =
-      GPR_MAX(backoff->current_timeout_millis, backoff->min_timeout_millis);
-  return grpc_exec_ctx_now(exec_ctx) + first_timeout;
+  backoff->current_backoff = backoff->initial_backoff;
+  const grpc_millis current_deadline =
```

Comment:
As we discussed, the backoff code needs to generate two values, `current_backoff` and `current_deadline`, where the latter may be greater than the former.  However, both this method and `grpc_backoff_step()` return only a single value, which it looks like is supposed to be `current_deadline`.  This seems fundamentally wrong, because the calling code needs to know and use both values.  So I think we need some sort of API change here.

In particular, the subchannel connection code uses the return value for two things:
- Scheduling an alarm for the next attempt.
- The deadline for the current attempt.

We need to return two values so that the subchannel connection code knows the right value to use for both of these things.

There are three other callers of this code as well: the grpclb LB policy and both DNS resolver implementations.  They all seem to use the return value just to schedule a timer for the next attempt; they don't seem to use it for setting deadlines at all, which may actually skew their use of this code -- the algorithm in the spec says that they need to take into account the time taken by each attempt, but they're not doing that.  So they may need some changes as well.  (Although the spec is really about connection backoff, so it would also be fine to say that these other cases don't really need to behave exactly the same way -- we should just make sure that we're happy with whatever behavior we are implementing here.  If we find that we aren't actually happy with the same behavior, it would be fine to use different code for these cases than for the subchannel connection code.)

Reply:
@markdroth I do not agree that `current_backoff` should be returned. `current_backoff` should be an internal variable that is rolled over in each backoff step internally in backoff.cc. Callers do not need to know about `current_backoff` at all when making TryConnect() (see that in [spec](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md)).

What callers should know is `min_connect_timeout`. They should make a connect with time-out being `MAX(current_deadline, now + min_connect_timeout)`. If they failed after `current_deadline', they immediately call `backoff_step` to get a new `current_backoff` and attempt another connection immediately. If they failed before `current_deadline`, they wait until `current_deadline` then do the same above.

Reply:
The problem with calling "step" again in this fashion is that the multiplier will be applied to the value of `current_backoff`, whereas if both values are returned at once, we don't incur in the side-effects (namely applying multiplier) of stepping.